### PR TITLE
idl: add a Position struct to wrap reported lines

### PR DESCRIPTION
**Pull request by jparise**
Line-based data types are too limiting should we add support for other
positional information, such as offsets/columns (#349).

### CODE REVIEW

**Comment by jparise** [comment]
This relates to both #349 and #493. I'd like to get (something like) this in place before the line-based interface exported by #492 makes it into a published release.

**Comment by jparise** [comment]
A few questions:

- Is `Position` or `Location` the right word? I think either is fine, but now is the time to choose.
- Should `Position` be embedded in the `PositionError` struct or be a distinct `Pos` field?

**Comment by abhinav** [comment]
Thanks for the PR!

> Is Position or Location the right word? I think either is fine, but now is the time to choose.

Agreed. Either is fine by me. I think location reads nicer than position in this context, but there's [prior art on position representing locations](https://golang.org/pkg/go/token/#Position). I'll defer this choice to you.

> Should Position be embedded in the PositionError structs or be a distinct Pos field?

I think I'd prefer a Pos field (or Loc if we name the type Location).

**Comment by jparise** [comment]
> > Is Position or Location the right word? I think either is fine, but now is the time to choose.
> 
> Agreed. Either is fine by me. I think location reads nicer than position in this context, but there's [prior art on position representing locations](https://golang.org/pkg/go/token/#Position). I'll defer this choice to you.

I looked at some prior art. The [Language Server Protocol (LSP) specification](https://microsoft.github.io/language-server-protocol/specifications/specification-current) uses both words with specific meanings:

- [`Position`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#position) is essential a (line, offset) pair, like a coordinate.
- [`Location`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#location) is a higher-level representation that can include a range of positions (start, end) as well as a URI (such as a filename).

Given the buffer-based design of the IDL module, we can't provide a URI (filename), although we could provide a way to inject one such as via the `idl.Config` + `ast.Info` concepts proposed in #493. We could also provide ranges as an advanced implementation of #349.

So I think `Position` is accurate for what we're probably going to be able to provide in the near term, but `Location` would buy is more room to maneuver in the future. The complication is that by introducing `Location` now, we'd be doing so knowing that we'd want to extend it in the future, because it might be too limiting to ship:

```go
type Location struct {
    Line int
    // ... and then add:
    Offset int
    // ... and then realize we want (start, end) ranges
}
```

All that being said, I think we could commit to doing the following before we release a new version (so we have the flexibility to iterate a little bit within the release window):

1. Implement #349, and as part of that implementation, decide if we'll support ranges or just (line, offset).
2. Based on (1), we'll know the general shape of the `Location` struct.
3. Choose whether to also add a `Filename` field to the `Location` struct as well. (I don't think this is especially useful unless we offer recursive document parsing, though.)

My current instinct is that we'll probably _not_ implement ranges in (1) or filenames in (3), and then we'll end up with:

```go
type Location struct {
    Line int
    Offset int
}
```

... which is effectively just a `Position`.

Not to just turn this back around at you, but do you land in the same place?

**Comment by abhinav** [comment]
Those are valid points. Yeah, I suspect we won't end up doing ranges,
in which case Position is fine as-is.

(Although, naming nit: `Offset`, if we add it, would be byte offset
within a file. The column number in a line would be on a `Column`
field.)

---

If we really want to track ranges, there's the possibility of an
alternative, slightly more complicated design inspired by go/token and
go/ast. The gist of it is:

- track position ranges as byte offsets in the file `(start, end)`
- expose them via `Info.Start(Node)` and `Info.End(Node)` methods
  returning offsets, similar to how we were going to do line numbers
- separately in Info, track a list of byte offsets in the file at which
  newlines occur (a sorted `[]Offset`)
- `Info.Position(Offset) Position` returns a `Position{Line, Column}`,
  computed dynamically from the knowledge of newline offsets
  (approximately: binary search in `[]Offset` to get line number,
  subtract offset for that line to get column number).
  
So the rough surface for this is:

```go
type Offset int
// To make the zero offset mean "unknown" (to account for existing code
// and struct literals), we can make offset 1-indexed in the file.

type Position struct {
    Line int // 1-indexed line number
    // In the future,
    //   File string  // if known
    //   Column int
}

type Info struct {
    newlines []Offset
    otherStarts map[ast.Node]Offset
}

func (*Info) Start(Node) Offset

// In the future,
//   func (*Info) End(Node) Offset

func (*Info) Position(Offset) Position

// Optional convenience methods:
//   func (*Info) StartPosition(Node) Position{ i.Position(i.Start(n)) }
//   func (*Info) EndPosition(Node) Position{ i.Position(i.End(n)) }

type Config struct {
    Info *Info

    // In the future,
    //  File string // name of the file, if known
}

func (*Config) Parse([]byte) (*ast.Program, error)
```

Used like so:

```go
var info Info
cfg := idl.Config{Info: &info, File: "foo.thrift"}

prog, err := cfg.Parse(..)
if err != nil {
  // ...
}

n := someNode(prog)

startPos := info.Position(info.Start(n))
fmt.Println(startPos.Line)  // 42

// In the future:
//   fmt.Println(startPos)  // foo.thrift:42:5
//
// Also in the future:
//   fmt.Println(info.EndPosition(n))  // foo.thrift:45:1
```

This is similar to how `go/token.FileSet` and `go/token.Pos` work to
track positions of Go AST nodes.

Advantages: this can support column numbers, file names, and ranges,
and costs only one integer per node (two if/when we add end positions).

Disadvantages: it's a lot more API surface area, it's more complicated
to implement and use, and it leaves a redundant `Line` field on all
existing AST nodes that have it.

---

I'm throwing the second design idea mostly as a maybe. Right now, I
don't have the time to implement it, and we don't have the need for
that functionality. I don't want to volunteer your time for anything
you don't need from this library.

So, if you really want column tracking and positional ranges, we'd
welcome an implementation of the above, or of an alternative/better
proposal.

Equally valid: if you don't have need of this functionality in a
timeline that matters, I'm also okay with a simple `Position` struct
that leaves room for a Column field in the future.

**Comment by jparise** [comment]
I think I'm good sticking with what's in this change for now. I'm motivated to get these two things done next:

1. Line numbers from #493 (including the proposed `Config` concept)
2. Column numbers from #349

... but I don't have a need for ranges at this point.

File: idl/error.go [hunk 1]
```diff
@@ -29,30 +29,33 @@ import (
...
-type ParseError struct{ Lines []LineError }
+type ParseError struct{ Errors []Error }
 
-// LineError holds an error and the line number that caused it.
-type LineError struct {
-	Line int
-	Err  error
+// Error holds an error and the position that caused it.
+type Error struct {
+	Pos Position
```
Comment (abhinav) [inline idl/error.go:36]: Note to other reviewers: Not a breaking change. We haven't released this yet.

**Review by abhinav — APPROVED** [review]
Sounds good, thanks!

**Pull request merged**